pkg/stream/connpool/msgconnpool/connpool.go

```diff
@@ -0,0 +1,459 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package msgconnpool
+
+import (
+	"context"
+	"errors"
+	"math"
+	"sync"
+	"sync/atomic"
+	"time"
+	"unsafe"
+
+	v2 "mosn.io/mosn/pkg/config/v2"
+	"mosn.io/mosn/pkg/upstream/cluster"
+
+	"mosn.io/api"
+	"mosn.io/mosn/pkg/log"
+	"mosn.io/mosn/pkg/types"
+	"mosn.io/pkg/buffer"
+	"mosn.io/pkg/utils"
+)
+
+// for reconnect
+const (
+	notConnecting = iota
+	connecting
+)
+
+// types.ConnectionPool
+type connpool struct {
+	idleClients []*activeClient
+
+	host      types.Host
+	clientMux sync.Mutex
+
+	autoReconnectWhenClose bool
+	heartBeatCreator       func() KeepAlive
+	reconnTryTimes         int
+	readFilters            []api.ReadFilter
+
+	destroyed uint64
+}
+
+func (p *connpool) Host() types.Host {
+	return p.host
+}
+
+// GetActiveClient get a avail client
+func (p *connpool) GetActiveClient(ctx context.Context) (*activeClient, types.PoolFailureReason) {
+	p.clientMux.Lock()
+	defer p.clientMux.Unlock()
+
+	n := len(p.idleClients)
+
+	// no available client
+	if n == 0 {
+		c, reason := p.newActiveClientLocked(ctx)
+		// if the reason is connection failure
+		// will automatic reconnect, so we can save the active client
+		// to avoid concurrent reconnect
+		if c != nil {
+			// should put this conn to pool
+			p.idleClients = append(p.idleClients, c)
+		}
+
+		return c, reason
+	} else {
+		var lastIdx = n - 1
+		var reason types.PoolFailureReason
+		c := p.idleClients[lastIdx]
+		return c, reason
+	}
+}
+
+// Destroy the pool
+func (p *connpool) Destroy() {
+	atomic.StoreUint64(&p.destroyed, 1)
+
+	p.clientMux.Lock()
+	defer p.clientMux.Unlock()
+
+	for _, c := range p.idleClients {
+		c.host.Connection.Close(api.NoFlush, api.LocalClose)
```

Comment:
maybe panic if connection not connected?

Reply:
```go
func destroyWhenNotConnected() {
	server := tcpServer{
		addr : fmt.Sprint("127.0.0.1:10002"),
		//addr : fmt.Sprint("101.200.197.139:80"),
	}
	server.start()

	pool := NewConn(server.addr, 100000, func() KeepAlive { return nil }, nil, true)
	time.Sleep(time.Second)
	server.stop()
	time.Sleep(time.Second)
	pool.Destroy()
	println("not panic")
}
```
will not panic here